image-hd.c

```diff
@@ -721,19 +758,18 @@ static int hdimage_setup(struct image *image, cfg_t *cfg)
 			now = part->offset + part->size;
 	}
 
-	if (hd->gpt)
-		now += GPT_SECTORS * 512;
-
 	if (image->size > 0 && now > image->size) {
```

Comment:
This looks wrong. We need to check if the image is large enough after adding the gpt sectors.
Maybe move the check after the `if (image->size == 0) {` block?

Reply:
No, it's not wrong, "now" has implicitly been bumped when the loop over all partitions handled the fake partitions - which in the case of a pre-set image size includes the backup table. So "normally" (i.e. when the config is well-formed), now will then be exactly image->size.

If, prior to this change, some ordinary partition would have extended into where the backup table should live, the "now += GPT_SECTORS*512" would have bumped now beyond image size. But that case is now taken care of by noticing that the partition overlaps with the backup table. If, prior to this change, some partition started _beyond_ image->size (so it doesn't collide with the backup table), that partition's endpoint itself will have caused now to be beyond image->size.

"now" is max(partition_end(p) for p in image->partitions) including the "fake" ones, so the right thing to check is that none of the areas we know we're going to populate extend beyond image->size. I'm clearly biased, but I think the code is easier to read and reason about this way.

Reply:
I see. Maybe do this:
- always create the fake partition for the backup and set the offset to zero if the image size is not specified.
- keep the `gpt_backup` pointer, then you can do `partsize -= gpt_backup->size` later for the autoresize partition.
- add a special case to set the offset for `part == gpt_backup` to align the end to 4k (it has an odd number of sectors).

Reply:
I guess I can do (something like) that. Out of curiosity, where does the "must end on a 4K boundary" come from? It's done in the "image size not specified" case, but it's not enforced when the image size is given explicitly, and I don't know anywhere that mandates this.

Reply:
OK, so I pushed the last update I'm gonna be able to do this side of Easter.

I changed the hole syntax to be the `(5K; 7M)` as discussed, added documentation for it (in a separate patch, easier that way if implementation or documentation should change, feel free to squash). 

I also added a final patch treating gpt_backup along the lines you suggested. Rebasing/squashing that one into the commit introducing the fake partitions is a little tedious, but I can do it if that's desirable.

Thanks so far for all your careful review and suggestions!